### PR TITLE
Use `upvar_tys` in more places, make it return a list

### PR DESCRIPTION
**Pull request by compiler-errors**
Just a cleanup that fell out of a PR that I was gonna write, but that PR kinda got stuck.

### CODE REVIEW

**Comment by compiler-errors** [comment]
probably shouldnt matter, but let's see

@bors try @rust-timer queue

File: compiler/rustc_middle/src/ty/layout.rs [hunk 1]
```diff
@@ -906,7 +906,7 @@ where
                         if i == tag_field {
                             return TyMaybeWithLayout::TyAndLayout(tag_layout(tag));
                         }
-                        TyMaybeWithLayout::Ty(args.as_generator().prefix_tys().nth(i).unwrap())
+                        TyMaybeWithLayout::Ty(*args.as_generator().prefix_tys().get(i).unwrap())
```
Comment (oli-obk) [inline compiler/rustc_middle/src/ty/layout.rs:909]:
```suggestion
                        TyMaybeWithLayout::Ty(args.as_generator().prefix_tys()[i])
```

File: compiler/rustc_mir_dataflow/src/elaborate_drops.rs [hunk 2]
```diff
@@ -861,7 +861,7 @@ where
         let ty = self.place_ty(self.place);
         match ty.kind() {
             ty::Closure(_, args) => {
-                let tys: Vec<_> = args.as_closure().upvar_tys().collect();
+                let tys: Vec<_> = args.as_closure().upvar_tys().to_vec();
                 self.open_drop_for_tuple(&tys)
```
Comment (oli-obk) [inline compiler/rustc_mir_dataflow/src/elaborate_drops.rs:865]: is this allocation necessary? you can convert a list to a slice directly

File: compiler/rustc_mir_dataflow/src/elaborate_drops.rs [hunk 3]
```diff
@@ -871,7 +871,7 @@ where
             // It effectively only contains upvars until the generator transformation runs.
             // See librustc_body/transform/generator.rs for more details.
             ty::Generator(_, args, _) => {
-                let tys: Vec<_> = args.as_generator().upvar_tys().collect();
+                let tys: Vec<_> = args.as_generator().upvar_tys().to_vec();
                 self.open_drop_for_tuple(&tys)
```
Comment (oli-obk) [inline compiler/rustc_mir_dataflow/src/elaborate_drops.rs:875]: same

File: compiler/rustc_mir_transform/src/generator.rs [hunk 4]
```diff
@@ -1471,7 +1471,7 @@ impl<'tcx> MirPass<'tcx> for StateTransform {
                 let args = args.as_generator();
                 (
                     args.discr_ty(tcx),
-                    args.upvar_tys().collect::<Vec<_>>(),
+                    args.upvar_tys().to_vec(),
```
Comment (oli-obk) [inline compiler/rustc_mir_transform/src/generator.rs:1474]: also possibly unnecessary alloc

File: compiler/rustc_trait_selection/src/traits/query/dropck_outlives.rs [hunk 5]
```diff
@@ -291,9 +291,9 @@ pub fn dtorck_constraint_for_ty_inner<'tcx>(
                 return Err(NoSolution);
             }
 
-            constraints.outlives.extend(
-                args.as_generator().upvar_tys().map(|t| -> ty::GenericArg<'tcx> { t.into() }),
-            );
+            constraints
+                .outlives
+                .extend(args.as_generator().upvar_tys().iter().map(|t| ty::GenericArg::from(t)));
```
Comment (oli-obk) [inline compiler/rustc_trait_selection/src/traits/query/dropck_outlives.rs:296]:
```suggestion
                .extend(args.as_generator().upvar_tys().iter().map(ty::GenericArg::from));
```

File: compiler/rustc_middle/src/ty/layout.rs (hunk 1 above)
Comment (compiler-errors) [reply compiler/rustc_middle/src/ty/layout.rs:909]: ty::Lists are not indexable

Comment (oli-obk) [reply compiler/rustc_middle/src/ty/layout.rs:909]: huh. I thought their `Deref<Target = [T]>` impl would make that "just work"

Comment (compiler-errors) [reply compiler/rustc_middle/src/ty/layout.rs:909]: ok i lied, for some reason it works now

**Pull request merged**